renku/core/management/datasets.py

```diff
@@ -356,30 +388,56 @@ def _add_from_local(self, dataset, path, link, destination):
             'parent': self
         }]
 
-    def _add_from_url(self, dataset, url, destination):
+    def _add_from_urls(self, dataset, urls, destination, extract, interactive):
+        destination.mkdir(parents=True, exist_ok=True)
+
+        files = []
+        max_workers = min(os.cpu_count() + 4, 8)
+        with concurrent.futures.ThreadPoolExecutor(max_workers) as executor:
+            futures = {
+                executor.submit(
+                    self._add_from_url, dataset, url, destination, extract,
```

Comment:
This will lead to a regession where tqdm progress bars are broken (display bug) when importing datasets with many files (e.g. test with `doi:10.7910/DVN/TJCLKP`). the `id_queue` code higher up was added to fix that issue. The code used to create a fixed number of slots that progress bars can be shown in according to the number of concurrent processes and each worker would only write to the progress bar in its slow, preventing multiple parallel workers with progress bars from interfering with each other.

Reply:
I've tested this before and I've tested it again with DOI you provided and the progress bars do not overlap and do not leave a trace. The difference is that this new implementation uses multiple threads instead of multiple processes. TQDM takes care of the overlaps in the multi-threaded mode and we don't need to do anything special. You can test it with something that takes a while to download and see the progress bars next to each other: `https://doi.org/10.5281/zenodo.3583477`

Moreover, the current implementation shows only one progress bar at any time; I did not check that if it is just the display or it does not download in parallel in the first place. If you checkout the master branch and use the above DOI you should be able to re-produce this issue. I've found this when I was doing the refactoring and since this PR solves that issue I did not create a bug report.

Reply:
The problem comes when there's less lines in the terminal than files being downloaded, then it gets a weird scrolling behaviour/display bug. You can see it if you drag the terminal smaller than fullscreen. With the current master there's 4 bars at a time (depends in number of CPUs), which doesn't have this issue. So I don't see the issue you mention on master:
![master_progress](https://user-images.githubusercontent.com/664486/71871900-10906d00-311b-11ea-8a06-a366f90702e0.png)


Not sure if a cosmetic issue like this is worth our time, though, it's a bit complicated :)

Reply:
You are right in both cases. Sorry for my comment about the current implementation; for some reason I was excepting to see 5 parallel downloads per core ;).
Please let me know what do you think we should do regarding this corner case so that we can close this PR: keep the new code or revert to the old one? I'd prefer to keep the new code since it is more concise. Someone who is using less than 8 lines in console probably doesn't care much about polluted console output :).

Reply:
I also think it's not worth fixing. The fix is uglier than the problem, it's purely cosmetic and only happens very rarely. With the old implementation it wasn't too big of a change to fix it, but now that it's combined with renku add, I think we should just leave it this way and just live with the display issue :).